### PR TITLE
Support building cerebral on non-posix environments

### PR DESCRIPTION
**Pull request by reflog**

### CODE REVIEW

**Comment by reflog** [comment]
The travis job is failing, I'm not sure why :( I've tested that it's working both on Linux and Windows

**Comment by henri-hulski** [comment]
Try to add cross-env to the dev-dependencies of the monorepo.

**Comment by reflog** [comment]
same result

**Comment by henri-hulski** [comment]
Hmm in the Travis log it says 
```
Error: Command failed: npm run prepublish 
sh: 1: cross-env: not found
```

**Comment by henri-hulski** [comment]
@reflog have a look at #365.
There @fopsdev uses
```js
   "build": "node ../../node_modules/cross-env/bin/cross-env.js BABEL_ENV=production ../../node_modules/.bin/babel src/ --out-dir=lib/ -s --presets es2015",
```

**Comment by henri-hulski** [comment]
So it seems that you have to use the path to the bin directory of the monorepo node_modules to run cross-env.

**Comment by Guria** [comment]
see a bin a section of monorepo package.json

File: package.json [hunk 1]
```diff
@@ -74,5 +62,9 @@
       "**/node_modules/**",
       "**/build/"
     ]
+  },
+  "devDependencies": {
```
Comment (Guria) [inline package.json]: move to general deps for consistency

**Comment by Guria** [comment]
hmm. seems link-parent-bin does the same as my trick with bin section

**Comment by henri-hulski** [comment]
Now there is a different error 
```
> function-tree@0.5.0-b build /home/travis/build/cerebral/cerebral/packages/function-tree
> cross-env BABEL_ENV=production babel src/ --out-dir=lib/ -s
Errored while running BootstrapCommand.execute
Error: Command failed: npm run prepublish 
events.js:160
      throw er; // Unhandled 'error' event
      ^
Error: spawn babel ENOENT
    at exports._errnoException (util.js:1018:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:193:32)
    at onErrorNT (internal/child_process.js:367:16)
    at _combinedTickCallback (internal/process/next_tick.js:80:11)
    at process._tickCallback (internal/process/next_tick.js:104:9)
    at Module.runMain (module.js:606:11)
    at run (bootstrap_node.js:393:7)
    at startup (bootstrap_node.js:150:9)
    at bootstrap_node.js:508:3
```

**Comment by reflog** [comment]
@guria I've used link-parent-bin instead of your trick, because on Windows you need to link to cmd files and not the ones without extension

**Comment by henri-hulski** [comment]
So it seems he finds now cross-env.

**Comment by reflog** [comment]
ok, fixed the issues with CI, should be all good now. less hacks than before!

**Comment by Guria** [comment]
My bin trick uses standard nodejs mechanism and must be working in windows as well. You just need add an appropriate entry into it.

I suppose we should keep only one of:
* my bin trick whith monorepo being controlled by lerna
* use link-parent-bin package
* use new lerna [hoisting mechanism](https://github.com/lerna/lerna/blob/master/doc/hoist.md) 

All of them has own pros and cons which I'd like to discuss. I will write up a bit more details on it this weekend I hope.

File: package.json [hunk 2]
```diff
@@ -4,9 +4,10 @@
   "version": "0.0.1",
   "license": "MIT",
   "scripts": {
-    "setup": "lerna bootstrap && link-parent-bin -s true",
-    "setup:packages": "lerna bootstrap --scope \"@(function-tree|cerebral*)\" && link-parent-bin -s true",
-    "setup:babel": "lerna bootstrap --scope \"@(function-tree|cerebral*|babel*)\" && link-parent-bin -s true",
+    "link": "link-parent-bin -s true && ink-parent-bin -s true -c babel",
```
Comment (henri-hulski) [inline package.json]: `ink-parent-bin`?

**Comment by reflog** [comment]
@Guria - unfortunately bin trick does not work on windows. It links the wrong script.
@henri-hulski - gosh, no idea how that slipped by. will fix

File: package.json [hunk 3]
```diff
@@ -4,9 +4,10 @@
   "version": "0.0.1",
   "license": "MIT",
   "scripts": {
-    "setup": "lerna bootstrap",
-    "setup:packages": "lerna bootstrap --scope '@(function-tree|cerebral*)'",
-    "setup:babel": "lerna bootstrap --scope '@(function-tree|cerebral*|babel*)'",
+    "link": "link-parent-bin -s true && ink-parent-bin -s true -c babel",
```
Comment (henri-hulski) [inline package.json]: I think I should comment here. 
`ink-parent-bin` seems to be a typo.

**Comment by henri-hulski** [comment]
@Guria the lerna hoisted approach looks cool. 
Nice that local package.json could stay unchanged. 
Don't know if modules resolution is a problem for us.

**Comment by Guria** [comment]
@reflog can you check one step more then? It seems there is no more need to monorepo being controlled by lerna. Can you try to remove it from lerna.json and from packages dev dependencies?

File: .travis.yml [hunk 4]
```diff
@@ -8,6 +8,8 @@ notifications:
   email: false
 before_script:
 - npm prune
+- link-parent-bin -s true
+- link-parent-bin -s true -c babel
```
Comment (Guria) [inline .travis.yml]: demos and website also must be linked for publishing purposes. and btw why not using npm run link here?

File: package.json [hunk 5]
```diff
@@ -15,19 +16,8 @@
...
-    "nyc": "./node_modules/.bin/nyc",
-    "react-scripts": "./node_modules/.bin/react-scripts",
-    "npm-run-all": "./node_modules/.bin/npm-run-all",
-    "run-p": "./node_modules/.bin/run-p",
-    "run-s": "./node_modules/.bin/run-s",
-    "shx": "./node_modules/.bin/shx",
-    "gh-pages": "./node_modules/.bin/gh-pages",
-    "jest": "./node_modules/.bin/jest"
+    "deploy": "cd website/publisher && npm run ci",
+    "link-parent-bin": "link-parent-bin"
```
Comment (Guria) [inline package.json]: is it extra now?

File: package.json [hunk 6]
```diff
@@ -4,9 +4,10 @@
   "version": "0.0.1",
   "license": "MIT",
   "scripts": {
-    "setup": "lerna bootstrap",
-    "setup:packages": "lerna bootstrap --scope '@(function-tree|cerebral*)'",
-    "setup:babel": "lerna bootstrap --scope '@(function-tree|cerebral*|babel*)'",
+    "link": "link-parent-bin -s true && link-parent-bin -s true -c babel",
+    "setup": "npm run link && lerna bootstrap",
```
Comment (Guria) [inline package.json]: maybe just call `npm run link` as postinstall script?

File: package.json (hunk 5 above)
Comment (reflog) [inline package.json]: yep. will remove

File: package.json (hunk 6 above)
Comment (reflog) [inline package.json]: that would make sense!

File: .travis.yml (hunk 4 above)
Comment (reflog) [inline .travis.yml]: will do

**Comment by reflog** [comment]
@Guria - can you elaborate what exactly do you want me to try here:
"can you check one step more then? It seems there is no more need to monorepo being controlled by lerna. Can you try to remove it from lerna.json and from packages dev dependencies?"

**Comment by Guria** [comment]
@reflog lerna.json has `"."` entry that instructs lerna to treat monorepo as dependency. It was necessary to make bin trick working. And seems there is no need in it anymore. And after that we should remove @cerebral/monorepo from dev deps across packages.

File: package.json [hunk 7]
```diff
@@ -4,9 +4,10 @@
   "version": "0.0.1",
   "license": "MIT",
   "scripts": {
-    "setup": "lerna bootstrap",
-    "setup:packages": "lerna bootstrap --scope '@(function-tree|cerebral*)'",
-    "setup:babel": "lerna bootstrap --scope '@(function-tree|cerebral*|babel*)'",
+    "link": "link-parent-bin -s true && link-parent-bin -s true -c babel && link-parent-bin -s true -c demos && link-parent-bin -s true -c website",
+    "setup": "npm run link && lerna bootstrap",
```
Comment (Guria) [inline package.json]: I would suggest to remove link from here, since now we have it in postinstall

**Comment by reflog** [comment]
removing the '.' from package didn't work. broke the build :(

**Comment by Guria** [comment]
Just removing dot is not enough. I have mentioned that extra dependency on monorepo across packages should be removed.

**Comment by reflog** [comment]
all done. works like a charm

**Pull request merged**